### DISCUSSION TITLE
Google软件工程之过程篇 · BMPI

### DISCUSSION THREAD

**Discussion by giscus** [bmpi-dev/bmpi.dev#145] [Announcements]
# Google软件工程之过程篇 · BMPI

本文是《Software Engineering at Google》的读书笔记，同时会穿插分享我对软件工程的理解。本文主要介绍软件工程的过程，主要包括Code Review、技术文档与自动化测试（单元、集成、E2E）。

https://www.bmpi.dev/dev/software-engineering-at-google/process/

**Comment by madawei2699** [comment]
## 关于测试覆盖率的迷思

百分百的测试覆盖率并不意味着高质量，因为无业务意义的单元测试是可以让测试覆盖率达到百分百的，但这些单元测试只是在用替身数据，这些替身数据或状态并不能代表业务含义，于是测试覆盖率达到百分百了，但软件却有很多不符合业务意义的错误。这种情况需要考虑加入更多集成测试或端对端的业务测试，这样从更大范围的视角去测试系统是否满足业务需求，如果业务有变动，那这些大的测试也需要满足业务需求的变动，从而进一步让软件开发人员能意识到需要更新相关的单元测试及替身数据。

**Reply by madawei2699** [reply]
背景是今天和同事结对开发，本来一个模块的测试覆盖率都达到百分百了，但新加入一点非常简单的代码却让好几个单元测试失败了，进一步分析发现 mock 的数据都是错误的，系统的功能符合业务需求，测试用了错误的测试数据，覆盖率还达到百分百🤔这些奇怪的事情让我们意识到百分百的测试覆盖率并不代表系统达到了高质量，虽然之前已经听到过这个说法，但今天的体会反而更深刻了。

**Reply by madawei2699** [reply]
https://cs-syd.eu/posts/2021-10-22-why-mocking-is-a-bad-idea
